### ISSUE TITLE
Pass the list of requested fieldnames to the data provider

### ISSUE THREAD

**Issue by TMiguelT** [marmelab/react-admin#3767]
## Problem
Currently, the data providers are not aware of which fields are actually going to be used for a request. For example, if we had the following form for comment 123:

```html
<SimpleShowLayout>
    <TextField source="id"/>
    <TextField source="content"/>
    <DateField source="created_at"/>
    <ReferenceField label="Owner" source="user_id" reference="users">
        <TextField source="username"/>
    </ReferenceField>
</SimpleShowLayout>
```

The call to the data provider would look like: `dataProvider(GET_ONE, 'posts', { id: 123 });`, meaning that a REST API will likely generate a request something like : `GET /posts/123`.

However, there are cases where it is advantageous for the API to know which fields are being requested:

* Optimizing the database query, ensuring we don't add joins or calculate fields that we don't need. For the example above, we aren't displaying the parent `post` data, so there's no need to query for and return it in this request
* Allowing us to *add* fields that shouldn't be defaults. In the example above, we might want to fetch the data for the `<ReferenceField>` without having to make a second request, doing something like `GET /posts/123?include=user` (according to the [JSON API spec](https://jsonapi.org/format/#fetching-includes))

## Solution
Expand the request data for all DataProvider calls to include a list of field names that are being requested. For example, when compared to the current request data [shown here](https://marmelab.com/react-admin/DataProviders.html#request-format), we would now have:

Type | Usage | Params format
-- | -- | --
GET_LIST | Search for resources | `{ pagination: { page: {int} , perPage: {int} }, sort: { field: {string}, order: {string} }, filter: {Object}, fields: {string[]} }`
GET_ONE | Read a single resource, by id | `{ id: {mixed}, fields: {string[]} }`
GET_MANY | Read a list of resource, by ids | `{ ids: {mixed[]}, fields: {string[]} }`
GET_MANY_REFERENCE | Read a list of resources related to another one | `{ target: {string}, id: {mixed}, pagination: { page: {int} , perPage: {int} }, sort: { field: {string}, order: {string} }, filter: {Object}, fields: {string[]} }`

**Comment by fzaninotto** [comment]
Hi, and thanks for your suggestion.

The `Show` controller fetches the `dataProvider` without knowing which fields are present in its grandchildren - and before these grandchildren are fetched. We don't want to inspect children before fetching (we've gone that path before with ng-admin, and it's a dead end), so the `Show` controller currently calls `GET_LIST` without specifying the fields it needs. 

Most of the time, if you need to add a join in the server-side, you probably need to fetch the related data lazily using a `Reference-` field or Input on the client side. That's how react-admin was designed, and it provides excellent performance out of the box thanks to this strategy.   

Besides, we recommend against using `include`-like directives in your REST routes, as react-admin already optimizes related resources fetching and has a sophisticated resource-based cache (think about embedding CSS in the HTML versus having an external, cacheable CSS file shared among HTML pages).

If all the requests for a given resource must specify the fields, you can already do so in the `dataProvider`.

To conclude, the only case we could handle would be specific controllers accepting a `fields` parameter and passing it to the `dataProvider`. This is outside of the scope we want to address in the react-admin core, and it's very feasible using the new react-admin hooks we're publishing in v3. We invite you to do it in userland. And if you ever want to make it an open-source package, feel free to add a link to it in the Ecosystem documentation by opening a Pull Request here.

**Comment by TMiguelT** [comment]
Hi, thanks for your comprehensive answer.

> If all the requests for a given resource must specify the fields, you can already do so in the dataProvider.

How can I do this if I have no way to access the fields in the `dataProvider`?

> To conclude, the only case we could handle would be specific controllers accepting a fields parameter and passing it to the dataProvider

This, I think, would also be a fine solution for my use case. I'll wait for it to be released in v3 before I investigate how easy this would be.

**Comment by fzaninotto** [comment]
>> If all the requests for a given resource must specify the fields, you can already do so in the dataProvider.

> How can I do this if I have no way to access the fields in the dataProvider?

Let me rephrase:  If all the requests for a given resource must specify the **same** fields, you can already do so in the dataProvider. The controller won't decide the fields to fetch, the dataProvider will.

**Comment by TMiguelT** [comment]
Oh right, you mean writing an application-specific `dataProvider` with a list of fields baked in. But this wouldn't work for a generalised `dataProvider` that works for any conformant API.

**Comment by Abenezer** [comment]
> Oh right, you mean writing an application-specific `dataProvider` with a list of fields baked in. But this wouldn't work for a generalised `dataProvider` that works for any conformant API.

Plus lest say you want different props for different instance of components, it needs to be passed to dataProvider 
eg in odata 
`<ReferenceInput expand="customers" ... >   
<ReferenceInput expand="users" ...>  `
by the way the readme mentions ODATA provider, but i can't see how to fully implement it without this extra props

**Comment by Abenezer** [comment]
[ok people are just passing custom props inside filter props ](https://stackoverflow.com/questions/55713128/react-admin-how-to-pass-more-parameters-to-a-dataprovider-for-type-get-list#)
`<ReferenceInput filter={{customProp:xx}}`   it works but it seems hacky solution. if this is ok can you add it to docs
